### PR TITLE
mprotect: add syscall

### PR DESCRIPTION
**Pull request by badochov**
JIRA: RTOS-666

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: vm/map.c [hunk 1]
```diff
@@ -719,6 +730,71 @@ int vm_munmap(vm_map_t *map, void *vaddr, size_t size)
...
+	int attrs = 0;
+	map_entry_t *e;
+
+	if ((((ptr_t)vaddr) & (SIZE_PAGE - 1)) != 0) {
+		return -EINVAL;
+	}
+
+	if ((prot & PROT_READ) != 0) {
+		attrs |= PGHD_PRESENT;
+	}
```
Comment (lukileczo) [inline vm/map.c:747]: Now that the GR712RC port is merged, please add `PGHD_READ` to `attrs` - on this target pages can be exec-only - https://github.com/phoenix-rtos/phoenix-rtos-kernel/commit/db4f5e5e7db0eb10af6e2386cee98b5c0dc89ed3

File: include/syscalls.h [hunk 2]
```diff
@@ -69,7 +69,7 @@
 	ID(signalMask) \
 	ID(signalSuspend) \
 	ID(priority) \
-	\
+\
```
Comment (agkaminski) [inline include/syscalls.h:72]: Why this change? Imho better to keep indentation level. Perhaps there are some objective arguments?

File: include/sysinfo.h [hunk 3]
```diff
@@ -47,6 +47,7 @@ typedef struct _entryinfo_t {
 
 	unsigned char flags;
 	unsigned char prot;
+	unsigned char origProt;
```
Comment (agkaminski) [inline include/sysinfo.h:50]: how about `protOrig` instead? Easier to spot that it refers to the same stuff

File: syscalls.c [hunk 4]
```diff
@@ -115,6 +115,24 @@ int syscalls_munmap(void *ustack)
...
+{
+	void *vaddr;
+	size_t len;
+	int prot, err;
+
+	GETFROMSTACK(ustack, void *, vaddr, 0);
+	GETFROMSTACK(ustack, size_t, len, 1);
+	GETFROMSTACK(ustack, int, prot, 2);
+
+	err = vm_mprotect(proc_current()->process->mapp, vaddr, len, PROT_USER | prot);
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply syscalls.c:128]: How about adding `mapp` and `pmapp` to .codespell_ignore?

File: vm/map.c [hunk 5]
```diff
@@ -719,6 +726,144 @@ int vm_munmap(vm_map_t *map, void *vaddr, size_t size)
...
+}
+
+
+static void vm_mapEntrySplit(process_t *p, vm_map_t *m, map_entry_t *e, map_entry_t *new, size_t len)
+{
+	vm_mapEntryCopy(new, e);
+
+	new->vaddr += len;
+	new->size -= len;
+	new->aoffs += e->size;
```
Comment (agkaminski) [inline vm/map.c:744]: Shouldn't be `len` instead of `e->size`?

File: vm/map.c [hunk 6]
```diff
@@ -719,6 +726,144 @@ int vm_munmap(vm_map_t *map, void *vaddr, size_t size)
...
+	int result = EOK;
+	void *currVaddr;
+	size_t lenLeft = len, currSize, needed;
+	process_t *p = proc_current()->process;
+	addr_t pa;
+	int attr;
+	map_entry_t *e, *buf = NULL, *prev;
+	map_entry_t t;
+
+	if (((((ptr_t)vaddr) & (SIZE_PAGE - 1)) != 0) || ((len & (SIZE_PAGE - 1)) != 0)) {
```
Comment (agkaminski) [inline vm/map.c:765]:
```suggestion
	if (((((ptr_t)vaddr) & (SIZE_PAGE - 1)) != 0) || (len == 0) || ((len & (SIZE_PAGE - 1)) != 0)) {
```

File: vm/map.c [hunk 7]
```diff
@@ -863,10 +1008,7 @@ int vm_mapCopy(process_t *proc, vm_map_t *dst, vm_map_t *src)
 			return -ENOMEM;
 		}
 
-		hal_memcpy(f, e, sizeof(map_entry_t));
-		f->amap = amap_ref(e->amap);
-		amap_getanons(f->amap, f->aoffs, f->size);
-		f->object = vm_objectRef(e->object);
+		vm_mapEntryCopy(f, e);
```
Comment (agkaminski) [inline vm/map.c:1011]: Shouldn't `_map_add` below be removed?

File: include/syscalls.h (hunk 2 above)
Comment (badochov) [reply include/syscalls.h:72]: Our sincerlys overloard clang-format' whims, I'll revert those

File: syscalls.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (badochov) [reply syscalls.c:128]: created PR

File: vm/map.c (hunk 7 above)
Comment (agkaminski) [reply vm/map.c:1008]: Ah, yes, I confused it with `split`

File: include/syscalls.h [hunk 8]
```diff
@@ -121,4 +123,7 @@
 	ID(release) \
 	ID(sbi_putchar) \
 	ID(sbi_getchar) \
-	ID(sigreturn)
+	ID(sigreturn) \
+	\
+	ID(mprotect)
+/* clang-format on */
```
Comment (agkaminski) [inline include/syscalls.h:129]: No new line at the end of the file

**Review by agkaminski — COMMENTED** [review]
Last nitpick, otherwise LGTM
